core/include/gnuradio-4.0/PmtCollections.hpp

```diff
@@ -0,0 +1,485 @@
+
+    [[nodiscard]] constexpr T& front() noexcept {
+        assert(_data != nullptr && _size > 0);
+        return _data[0];
+    }
+    [[nodiscard]] constexpr const T& front() const noexcept {
+        assert(_data != nullptr && _size > 0);
+        return _data[0];
+    }
+    [[nodiscard]] constexpr T& back() noexcept {
+        assert(_data != nullptr && _size > 0);
+        return _data[_size - 1];
+    }
+    [[nodiscard]] constexpr const T& back() const noexcept {
+        assert(_data != nullptr && _size > 0);
+        return _data[_size - 1];
+    }
+
+    void reserve(std::size_t new_cap) {
+        if (new_cap <= _capacity) {
+            return;
+        }
+        _grow_capacity(new_cap);
+    }
+
+    void shrink_to_fit() {
+        if (_size == _capacity) {
+            return;
+        }
+        if (_size == 0UZ) {
+            clear();
+            return;
+        }
+        _grow_capacity(_size);
+    }
+
+    template<class... Args>
+    T& emplace_back(Args&&... args) {
+        if (_size == _capacity) {
+            reserve(_next_cap(_capacity, _size + 1UZ));
+        }
+        T* where = _data + _size;
+        std::construct_at(where, std::forward<Args>(args)...);
+        ++_size;
+        return *where;
+    }
+    void push_back(const T& v) { (void)emplace_back(v); }
+    void push_back(T&& v) { (void)emplace_back(std::move(v)); }
+    void pop_back() noexcept {
+        if (_size == 0UZ) {
+            return;
+        }
+        if constexpr (!std::is_trivially_destructible_v<T>) {
+            std::destroy_at(_data + (_size - 1UZ));
+        }
+        --_size;
+    }
+
+    constexpr void resize(std::size_t new_size) {
+        if (new_size <= _size) { // shrink in place - no reallocation!
+            std::destroy_n(_data + new_size, _size - new_size);
+            _size = new_size;
+        } else if (new_size <= _capacity) { // grow within capacity - no reallocation!
+            std::uninitialized_value_construct_n(_data + _size, new_size - _size);
+            _size = new_size;
+        } else {                            // must reallocate
+            _grow_capacity(new_size * 2UZ); // factor 2 growth factor
+            std::uninitialized_value_construct_n(_data + _size, new_size - _size);
+            _size = new_size;
+        }
+    }
+
+    void assign(std::size_t count, const T& value) {
+        clear();
+        if (count > 0) {
+            reserve(count);
+            std::uninitialized_fill_n(_data, count, value);
+            _size = count;
+        }
+    }
+
+    template<class It, class End>
+    requires std::input_iterator<It> && std::sentinel_for<End, It>
+    void assign(It first, End last) {
+        const auto new_size = static_cast<std::size_t>(std::ranges::distance(first, last));
+        clear();
+        if (new_size > 0) {
+            _data     = static_cast<T*>(_resource->allocate(new_size * sizeof(T), alignof(T)));
```

Comment:
allocate and dealocate are called in quite a few places. Might be useful to move into an inlined dedicated function.